File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Tests/Api/BookstoreServiceNegativeTest.cs
using System;
using System.Net;
using FluentAssertions;
using NUnit.Extension.DependencyInjection;
using RekindleTaf.Controllers;
using RekindleTaf.Exceptions;
using RekindleTaf.Model;
using RekindleTaf.Utils;

namespace RekindleTaf.Tests.Api;
[DependencyInjectingTestFixture]
public class BookstoreServiceNegativeTest(IBookstoreController bookstoreController) : Base
{
    [Test]
    public void Verify_CannotCreateBookstore_WithEmptyValues()
    {
        var bookstore = new Bookstore{Name = "", IsActive = null};
        var response = bookstoreController.Post(bookstore).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("name").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("isActive").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
    }
    
    [Test]
    public void Verify_CannotCreateBookstore_WithNullValues()
    {
        var bookstore = new Bookstore();
        var response = bookstoreController.Post(bookstore).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("name").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
        response.GetNode("isActive").Should().BeOneOf(ErrorMessage.NotBlank, ErrorMessage.NotNull);
    }
    
    [Test]
    public void Verify_CannotGet_NonExistingBookstore()
    {
        var response = bookstoreController.Get(Guid.NewGuid().ToString()).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.BookstoreNotFound);
    }
    
    [Test]
    public void Verify_CannotUpdate_NonExistingBookstore()
    {
        var bookstore = new Bookstore{Id = Guid.NewGuid().ToString(),Name = "Test", IsActive = true};
        var response = bookstoreController.Put(bookstore).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.BookstoreNotFound);
    }
    
    [Test]
    public void Verify_CannotDelete_NonExistingBookstore()
    {
        var bookstore = new Bookstore{Id = Guid.NewGuid().ToString()};
        var response = bookstoreController.Delete(bookstore).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.BookstoreNotFound);
    }
}

File: typescript-project/rekindle-book-store-aqa/tests/bookstore-service-negative.test.ts
import { beforeAll, describe, expect, test } from '@jest/globals';
import { Bookstore } from '@/model/domain-model';
import { BookstoreController } from '@/rest/controllers/bookstore-controller';
import { container } from 'tsyringe';
import { RestClient } from '@/rest/rekindle-rest-client';
import { HttpStatusCode } from 'axios';
import {ErrorMessage} from "../src/utils/utils";
//@ts-ignore
import {v4 as uuidv4} from 'uuid';

beforeAll(async () => {
    const restClient: RestClient = container.resolve("RestClient");
    await restClient.setupToken();
});

describe('Bookstore Service Negative Tests', () => {
    let bookstoreController: BookstoreController;

    beforeAll(() => {
        bookstoreController = container.resolve(BookstoreController);
    });

    test('Verify Cannot Create Bookstore With Empty Values', async () => {
        const bookstore: Bookstore = { name: '', isActive: null as any };
        try {
            const response = await bookstoreController.post(bookstore);
            expect(response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(response.data.name);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(response.data.isActive);
        } catch (error: any) {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.name);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.isActive);
        }
    });

    test('Verify Cannot Create Bookstore With Null Values', async () => {
        const bookstore: Bookstore = {};
        try {
            const response = await bookstoreController.post(bookstore);
            expect(response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(response.data.name);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(response.data.isActive);
        } catch (error: any) {
            expect(error.response.status).toBe(HttpStatusCode.BadRequest);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.name);
            expect([ErrorMessage.notBlank, ErrorMessage.notNull]).toContain(error.response.data.isActive);
        }
    });

    test('Verify Cannot Get Non-Existing Bookstore', async () => {
        try {
            const response = await bookstoreController.get(uuidv4());
            expect(response.status).toBe(HttpStatusCode.NotFound);
            expect(response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        } catch (error: any) {
            expect(error.response.status).toBe(HttpStatusCode.NotFound);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        }
    });

    test('Verify Cannot Update Non-Existing Bookstore', async () => {
        const bookstore: Bookstore = { id: uuidv4(), name: 'Test', isActive: true };
        try {
            const response = await bookstoreController.put(bookstore);
            expect(response.status).toBe(HttpStatusCode.NotFound);
            expect(response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        } catch (error: any) {
            expect(error.response.status).toBe(HttpStatusCode.NotFound);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        }
    });

    test('Verify Cannot Delete Non-Existing Bookstore', async () => {
        const bookstore: Bookstore = { id: uuidv4() };
        try {
            const response = await bookstoreController.delete(bookstore);
            expect(response.status).toBe(HttpStatusCode.NotFound);
            expect(response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        } catch (error: any) {
            expect(error.response.status).toBe(HttpStatusCode.NotFound);
            expect(error.response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
        }
    });
});